wgengine/netstack: data race in TestHandleLocalPackets

https://github.com/tailscale/tailscale/actions/runs/23556844837/job/68585893013?pr=19116

```
==================
WARNING: DATA RACE
Read at 0x000001f37078 by goroutine 738:
  tailscale.com/wgengine/netstack.maxInFlightConnectionAttemptsPerClient()
      /home/runner/work/tailscale/tailscale/src/wgengine/netstack/netstack.go:109 +0x27
  tailscale.com/wgengine/netstack.(*Impl).Start.(*Impl).wrapTCPProtocolHandler.func1()
      /home/runner/work/tailscale/tailscale/src/wgengine/netstack/netstack.go:549 +0x224
  gvisor.dev/gvisor/pkg/tcpip/stack.(*nic).DeliverTransportPacket()
      /home/runner/work/tailscale/tailscale/gomodcache/gvisor.dev/gvisor@v0.0.0-20260224225140-573d5e7127a8/pkg/tcpip/stack/nic.go:881 +0x576
  gvisor.dev/gvisor/pkg/tcpip/network/ipv4.(*endpoint).deliverPacketLocally()
      /home/runner/work/tailscale/tailscale/gomodcache/gvisor.dev/gvisor@v0.0.0-20260224225140-573d5e7127a8/pkg/tcpip/network/ipv4/ipv4.go:1394 +0x113e
  gvisor.dev/gvisor/pkg/tcpip/network/ipv4.(*endpoint).handleValidatedPacket()
      /home/runner/work/tailscale/tailscale/gomodcache/gvisor.dev/gvisor@v0.0.0-20260224225140-573d5e7127a8/pkg/tcpip/network/ipv4/ipv4.go:1208 +0xe24
  gvisor.dev/gvisor/pkg/tcpip/network/ipv4.(*endpoint).HandlePacket()
      /home/runner/work/tailscale/tailscale/gomodcache/gvisor.dev/gvisor@v0.0.0-20260224225140-573d5e7127a8/pkg/tcpip/network/ipv4/ipv4.go:931 +0x77e
  gvisor.dev/gvisor/pkg/tcpip/stack.(*nic).DeliverNetworkPacket()
      /home/runner/work/tailscale/tailscale/gomodcache/gvisor.dev/gvisor@v0.0.0-20260224225140-573d5e7127a8/pkg/tcpip/stack/nic.go:781 +0x20d
  tailscale.com/wgengine/netstack.(*linkEndpoint).DeliverLoopback()
      /home/runner/work/tailscale/tailscale/src/wgengine/netstack/link_endpoint.go:234 +0x2d4
  tailscale.com/wgengine/netstack.(*Impl).inject()
      /home/runner/work/tailscale/tailscale/src/wgengine/netstack/netstack.go:1072 +0x11c
  tailscale.com/wgengine/netstack.(*Impl).Start.gowrap1()
      /home/runner/work/tailscale/tailscale/src/wgengine/netstack/netstack.go:645 +0x2e

Previous write at 0x000001f37078 by goroutine 695:
  tailscale.com/tstest.Replace[go.shape.int].func1()
      /home/runner/work/tailscale/tailscale/src/tstest/tstest.go:33 +0x35
  testing.(*common).Cleanup.func1()
      /home/runner/.cache/tsgo/f4de14a515221e27c0d79446b423849a6546e3a6/src/testing/testing.go:1317 +0x168
  testing.(*common).runCleanup()
      /home/runner/.cache/tsgo/f4de14a515221e27c0d79446b423849a6546e3a6/src/testing/testing.go:1667 +0x225
  testing.tRunner.func2()
      /home/runner/.cache/tsgo/f4de14a515221e27c0d79446b423849a6546e3a6/src/testing/testing.go:2030 +0x4c
  runtime.deferreturn()
      /home/runner/.cache/tsgo/f4de14a515221e27c0d79446b423849a6546e3a6/src/runtime/panic.go:668 +0x5d
  testing.(*T).Run.gowrap1()
      /home/runner/.cache/tsgo/f4de14a515221e27c0d79446b423849a6546e3a6/src/testing/testing.go:2101 +0x38

Goroutine 738 (running) created at:
  tailscale.com/wgengine/netstack.(*Impl).Start()
      /home/runner/work/tailscale/tailscale/src/wgengine/netstack/netstack.go:645 +0x6bc
  tailscale.com/wgengine/netstack.makeNetstack()
      /home/runner/work/tailscale/tailscale/src/wgengine/netstack/netstack_test.go:142 +0x904
  tailscale.com/wgengine/netstack.TestTCPForwardLimits_PerClient()
      /home/runner/work/tailscale/tailscale/src/wgengine/netstack/netstack_test.go:632 +0xb7
  testing.tRunner()
      /home/runner/.cache/tsgo/f4de14a515221e27c0d79446b423849a6546e3a6/src/testing/testing.go:2036 +0x21c
  testing.(*T).Run.gowrap1()
      /home/runner/.cache/tsgo/f4de14a515221e27c0d79446b423849a6546e3a6/src/testing/testing.go:2101 +0x38

Goroutine 695 (finished) created at:
  testing.(*T).Run()
      /home/runner/.cache/tsgo/f4de14a515221e27c0d79446b423849a6546e3a6/src/testing/testing.go:2101 +0xb12
  testing.runTests.func1()
      /home/runner/.cache/tsgo/f4de14a515221e27c0d79446b423849a6546e3a6/src/testing/testing.go:2585 +0x85
  testing.tRunner()
      /home/runner/.cache/tsgo/f4de14a515221e27c0d79446b423849a6546e3a6/src/testing/testing.go:2036 +0x21c
  testing.runTests()
      /home/runner/.cache/tsgo/f4de14a515221e27c0d79446b423849a6546e3a6/src/testing/testing.go:2583 +0x9e9
  testing.(*M).Run()
      /home/runner/.cache/tsgo/f4de14a515221e27c0d79446b423849a6546e3a6/src/testing/testing.go:2443 +0xf4b
  main.main()
      _testmain.go:66 +0x164
==================
==================
WARNING: DATA RACE
Read at 0x000001f37070 by goroutine 738:
  tailscale.com/wgengine/netstack.maxInFlightConnectionAttempts()
      /home/runner/work/tailscale/tailscale/src/wgengine/netstack/netstack.go:83 +0x44
  tailscale.com/wgengine/netstack.maxInFlightConnectionAttemptsPerClient()
      /home/runner/work/tailscale/tailscale/src/wgengine/netstack/netstack.go:117 +0x56
  tailscale.com/wgengine/netstack.(*Impl).Start.(*Impl).wrapTCPProtocolHandler.func1()
      /home/runner/work/tailscale/tailscale/src/wgengine/netstack/netstack.go:549 +0x224
  gvisor.dev/gvisor/pkg/tcpip/stack.(*nic).DeliverTransportPacket()
      /home/runner/work/tailscale/tailscale/gomodcache/gvisor.dev/gvisor@v0.0.0-20260224225140-573d5e7127a8/pkg/tcpip/stack/nic.go:881 +0x576
  gvisor.dev/gvisor/pkg/tcpip/network/ipv4.(*endpoint).deliverPacketLocally()
      /home/runner/work/tailscale/tailscale/gomodcache/gvisor.dev/gvisor@v0.0.0-20260224225140-573d5e7127a8/pkg/tcpip/network/ipv4/ipv4.go:1394 +0x113e
  gvisor.dev/gvisor/pkg/tcpip/network/ipv4.(*endpoint).handleValidatedPacket()
      /home/runner/work/tailscale/tailscale/gomodcache/gvisor.dev/gvisor@v0.0.0-20260224225140-573d5e7127a8/pkg/tcpip/network/ipv4/ipv4.go:1208 +0xe24
  gvisor.dev/gvisor/pkg/tcpip/network/ipv4.(*endpoint).HandlePacket()
      /home/runner/work/tailscale/tailscale/gomodcache/gvisor.dev/gvisor@v0.0.0-20260224225140-573d5e7127a8/pkg/tcpip/network/ipv4/ipv4.go:931 +0x77e
  gvisor.dev/gvisor/pkg/tcpip/stack.(*nic).DeliverNetworkPacket()
      /home/runner/work/tailscale/tailscale/gomodcache/gvisor.dev/gvisor@v0.0.0-20260224225140-573d5e7127a8/pkg/tcpip/stack/nic.go:781 +0x20d
  tailscale.com/wgengine/netstack.(*linkEndpoint).DeliverLoopback()
      /home/runner/work/tailscale/tailscale/src/wgengine/netstack/link_endpoint.go:234 +0x2d4
  tailscale.com/wgengine/netstack.(*Impl).inject()
      /home/runner/work/tailscale/tailscale/src/wgengine/netstack/netstack.go:1072 +0x11c
  tailscale.com/wgengine/netstack.(*Impl).Start.gowrap1()
      /home/runner/work/tailscale/tailscale/src/wgengine/netstack/netstack.go:645 +0x2e

Previous write at 0x000001f37070 by goroutine 695:
  tailscale.com/tstest.Replace[go.shape.int].func1()
      /home/runner/work/tailscale/tailscale/src/tstest/tstest.go:33 +0x35
  testing.(*common).Cleanup.func1()
      /home/runner/.cache/tsgo/f4de14a515221e27c0d79446b423849a6546e3a6/src/testing/testing.go:1317 +0x168
  testing.(*common).runCleanup()
      /home/runner/.cache/tsgo/f4de14a515221e27c0d79446b423849a6546e3a6/src/testing/testing.go:1667 +0x225
  testing.tRunner.func2()
      /home/runner/.cache/tsgo/f4de14a515221e27c0d79446b423849a6546e3a6/src/testing/testing.go:2030 +0x4c
  runtime.deferreturn()
      /home/runner/.cache/tsgo/f4de14a515221e27c0d79446b423849a6546e3a6/src/runtime/panic.go:668 +0x5d
  testing.(*T).Run.gowrap1()
      /home/runner/.cache/tsgo/f4de14a515221e27c0d79446b423849a6546e3a6/src/testing/testing.go:2101 +0x38

Goroutine 738 (running) created at:
  tailscale.com/wgengine/netstack.(*Impl).Start()
      /home/runner/work/tailscale/tailscale/src/wgengine/netstack/netstack.go:645 +0x6bc
  tailscale.com/wgengine/netstack.makeNetstack()
      /home/runner/work/tailscale/tailscale/src/wgengine/netstack/netstack_test.go:142 +0x904
  tailscale.com/wgengine/netstack.TestTCPForwardLimits_PerClient()
      /home/runner/work/tailscale/tailscale/src/wgengine/netstack/netstack_test.go:632 +0xb7
  testing.tRunner()
      /home/runner/.cache/tsgo/f4de14a515221e27c0d79446b423849a6546e3a6/src/testing/testing.go:2036 +0x21c
  testing.(*T).Run.gowrap1()
      /home/runner/.cache/tsgo/f4de14a515221e27c0d79446b423849a6546e3a6/src/testing/testing.go:2101 +0x38

Goroutine 695 (finished) created at:
  testing.(*T).Run()
      /home/runner/.cache/tsgo/f4de14a515221e27c0d79446b423849a6546e3a6/src/testing/testing.go:2101 +0xb12
  testing.runTests.func1()
      /home/runner/.cache/tsgo/f4de14a515221e27c0d79446b423849a6546e3a6/src/testing/testing.go:2585 +0x85
  testing.tRunner()
      /home/runner/.cache/tsgo/f4de14a515221e27c0d79446b423849a6546e3a6/src/testing/testing.go:2036 +0x21c
  testing.runTests()
      /home/runner/.cache/tsgo/f4de14a515221e27c0d79446b423849a6546e3a6/src/testing/testing.go:2583 +0x9e9
  testing.(*M).Run()
      /home/runner/.cache/tsgo/f4de14a515221e27c0d79446b423849a6546e3a6/src/testing/testing.go:2443 +0xf4b
  main.main()
      _testmain.go:66 +0x164
==================
```

/cc @tailscale/dataplane

Another:
https://github.com/tailscale/tailscale/actions/runs/24363666577/job/71149682191?pr=19375

```
=== RUN   TestHandleLocalPackets
    userspace.go:350:     ... [v1] using fake (no-op) OS network configurator
    userspace.go:354:     ... [v1] using fake (no-op) DNS configurator
==================
WARNING: DATA RACE
Read at 0x000001f4a138 by goroutine 963:
  tailscale.com/wgengine/netstack.maxInFlightConnectionAttemptsPerClient()
      /home/runner/work/tailscale/tailscale/src/wgengine/netstack/netstack.go:109 +0x27
  tailscale.com/wgengine/netstack.(*Impl).Start.(*Impl).wrapTCPProtocolHandler.func1()
      /home/runner/work/tailscale/tailscale/src/wgengine/netstack/netstack.go:549 +0x224
  gvisor.dev/gvisor/pkg/tcpip/stack.(*nic).DeliverTransportPacket()
      /home/runner/work/tailscale/tailscale/gomodcache/gvisor.dev/gvisor@v0.0.0-20260224225140-573d5e7127a8/pkg/tcpip/stack/nic.go:881 +0x576
  gvisor.dev/gvisor/pkg/tcpip/network/ipv4.(*endpoint).deliverPacketLocally()
      /home/runner/work/tailscale/tailscale/gomodcache/gvisor.dev/gvisor@v0.0.0-20260224225140-573d5e7127a8/pkg/tcpip/network/ipv4/ipv4.go:1394 +0x113e
  gvisor.dev/gvisor/pkg/tcpip/network/ipv4.(*endpoint).handleValidatedPacket()
      /home/runner/work/tailscale/tailscale/gomodcache/gvisor.dev/gvisor@v0.0.0-20260224225140-573d5e7127a8/pkg/tcpip/network/ipv4/ipv4.go:1208 +0xe24
  gvisor.dev/gvisor/pkg/tcpip/network/ipv4.(*endpoint).HandlePacket()
      /home/runner/work/tailscale/tailscale/gomodcache/gvisor.dev/gvisor@v0.0.0-20260224225140-573d5e7127a8/pkg/tcpip/network/ipv4/ipv4.go:931 +0x77e
  gvisor.dev/gvisor/pkg/tcpip/stack.(*nic).DeliverNetworkPacket()
      /home/runner/work/tailscale/tailscale/gomodcache/gvisor.dev/gvisor@v0.0.0-20260224225140-573d5e7127a8/pkg/tcpip/stack/nic.go:781 +0x20d
  tailscale.com/wgengine/netstack.(*linkEndpoint).DeliverLoopback()
      /home/runner/work/tailscale/tailscale/src/wgengine/netstack/link_endpoint.go:234 +0x2d4
  tailscale.com/wgengine/netstack.(*Impl).inject()
      /home/runner/work/tailscale/tailscale/src/wgengine/netstack/netstack.go:1072 +0x11c
  tailscale.com/wgengine/netstack.(*Impl).Start.gowrap1()
      /home/runner/work/tailscale/tailscale/src/wgengine/netstack/netstack.go:645 +0x2e

Previous write at 0x000001f4a138 by goroutine 921:
  tailscale.com/tstest.Replace[go.shape.int].func1()
      /home/runner/work/tailscale/tailscale/src/tstest/tstest.go:33 +0x35
  testing.(*common).Cleanup.func1()
      /home/runner/.cache/tsgo/dfe2a5fd8ee2e68b08ce5ff259269f50ecadf2f4/src/testing/testing.go:1317 +0x168
  testing.(*common).runCleanup()
      /home/runner/.cache/tsgo/dfe2a5fd8ee2e68b08ce5ff259269f50ecadf2f4/src/testing/testing.go:1667 +0x225
  testing.tRunner.func2()
      /home/runner/.cache/tsgo/dfe2a5fd8ee2e68b08ce5ff259269f50ecadf2f4/src/testing/testing.go:2030 +0x4c
  runtime.deferreturn()
      /home/runner/.cache/tsgo/dfe2a5fd8ee2e68b08ce5ff259269f50ecadf2f4/src/runtime/panic.go:668 +0x5d
  testing.(*T).Run.gowrap1()
      /home/runner/.cache/tsgo/dfe2a5fd8ee2e68b08ce5ff259269f50ecadf2f4/src/testing/testing.go:2101 +0x38

Goroutine 963 (running) created at:
  tailscale.com/wgengine/netstack.(*Impl).Start()
      /home/runner/work/tailscale/tailscale/src/wgengine/netstack/netstack.go:645 +0x6bc
  tailscale.com/wgengine/netstack.makeNetstack()
      /home/runner/work/tailscale/tailscale/src/wgengine/netstack/netstack_test.go:142 +0x904
  tailscale.com/wgengine/netstack.TestTCPForwardLimits_PerClient()
      /home/runner/work/tailscale/tailscale/src/wgengine/netstack/netstack_test.go:820 +0xb7
  testing.tRunner()
      /home/runner/.cache/tsgo/dfe2a5fd8ee2e68b08ce5ff259269f50ecadf2f4/src/testing/testing.go:2036 +0x21c
  testing.(*T).Run.gowrap1()
      /home/runner/.cache/tsgo/dfe2a5fd8ee2e68b08ce5ff259269f50ecadf2f4/src/testing/testing.go:2030 +0x4c
  runtime.deferreturn()
      /home/runner/.cache/tsgo/dfe2a5fd8ee2e68b08ce5ff259269f50ecadf2f4/src/runtime/panic.go:668 +0x5d
  testing.(*T).Run.gowrap1()
      /home/runner/.cache/tsgo/dfe2a5fd8ee2e68b08ce5ff259269f50ecadf2f4/src/testing/testing.go:2101 +0x38

Goroutine 963 (running) created at:
  tailscale.com/wgengine/netstack.(*Impl).Start()
      /home/runner/work/tailscale/tailscale/src/wgengine/netstack/netstack.go:645 +0x6bc
  tailscale.com/wgengine/netstack.makeNetstack()
      /home/runner/work/tailscale/tailscale/src/wgengine/netstack/netstack_test.go:142 +0x904
  tailscale.com/wgengine/netstack.TestTCPForwardLimits_PerClient()
      /home/runner/work/tailscale/tailscale/src/wgengine/netstack/netstack_test.go:820 +0xb7
  testing.tRunner()
      /home/runner/.cache/tsgo/dfe2a5fd8ee2e68b08ce5ff259269f50ecadf2f4/src/testing/testing.go:2036 +0x21c
  testing.(*T).Run.gowrap1()
      /home/runner/.cache/tsgo/dfe2a5fd8ee2e68b08ce5ff259269f50ecadf2f4/src/testing/testing.go:2101 +0x38

Goroutine 921 (finished) created at:
  testing.(*T).Run()
      /home/runner/.cache/tsgo/dfe2a5fd8ee2e68b08ce5ff259269f50ecadf2f4/src/testing/testing.go:2101 +0xb12
  testing.runTests.func1()
      /home/runner/.cache/tsgo/dfe2a5fd8ee2e68b08ce5ff259269f50ecadf2f4/src/testing/testing.go:2585 +0x85
  testing.tRunner()
      /home/runner/.cache/tsgo/dfe2a5fd8ee2e68b08ce5ff259269f50ecadf2f4/src/testing/testing.go:2036 +0x21c
  testing.runTests()
      /home/runner/.cache/tsgo/dfe2a5fd8ee2e68b08ce5ff259269f50ecadf2f4/src/testing/testing.go:2583 +0x9e9
  testing.(*M).Run()
      /home/runner/.cache/tsgo/dfe2a5fd8ee2e68b08ce5ff259269f50ecadf2f4/src/testing/testing.go:2443 +0xf4b
  main.main()
      _testmain.go:66 +0x164
==================
    logger.go:105:     ... dns: using dns.noopManager
    userspace.go:440:     ... link state: interfaces.State{defaultRoute=eth0 ifs={docker0:[172.17.0.1/16] eth0:[10.1.0.191/20 llu6]} v4=true v6=false}
Warning: sock.go:3910:     ... magicsock: [warning] failed to force-set UDP read buffer size to 7340032: operation not permitted; using kernel default values (impacts throughput only)
Warning: sock.go:3910:     ... magicsock: [warning] failed to force-set UDP write buffer size to 7340032: operation not permitted; using kernel default values (impacts throughput only)
Warning: sock.go:3910:     ... magicsock: [warning] failed to force-set UDP read buffer size to 7340032: operation not permitted; using kernel default values (impacts throughput only)
Warning: sock.go:3910:     ... magicsock: [warning] failed to force-set UDP write buffer size to 7340032: operation not permitted; using kernel default values (impacts throughput only)
    peermtu.go:95:     ... [v1] magicsock: peermtu: peer MTU status is false
    magicsock.go:729:     ... magicsock: disco key = d:66e73b0ce85c3a62
    userspace.go:533:     ... Creating WireGuard device...
    userspace.go:573:     ... Bringing WireGuard device up...
    wglog.go:66:     ... wg: [v2] UDP bind has been updated
    wglog.go:66:     ... wg: [v2] Interface state was Down, requested Up, now Up
    userspace.go:577:     ... Bringing router up...
    router_fake.go:21:     ... [v1] warning: fakeRouter.Up: not implemented.
    userspace.go:586:     ... Clearing router settings...
    router_fake.go:26:     ... [v1] warning: fakeRouter.Set: not implemented.
    userspace.go:590:     ... Starting network monitor...
    userspace.go:638:     ... Engine created.
    wglog.go:66:     ... wg: [v2] Routine: receive incoming receiveDERP - started
    wglog.go:66:     ... wg: [v2] Routine: receive incoming mkReceiveFunc - started
    wglog.go:66:     ... wg: [v2] Routine: receive incoming mkReceiveFunc - started
    logger.go:105:     ... pm: migrating "_daemon" profile to new format
    envknob.go:84:     ... envknob: TS_DEBUG_NETSTACK="true"
    logpolicy.go:250:     ... logpolicy: using system state directory "/home/runner/.local/share/tailscale"
    local.go:1020:     ... linkChange: in state NoState; PAC or proxyConfig changed; updating routes
    local.go:2953:     ... [v1] netmap packet filter: (not ready yet)
    local.go:4734:     ... EditPrefs: MaskedPrefs{AdvertiseRoutes=[fd7a:115c:a1e0:b1a:0:7:a01:100/120] AdvertiseServices=["svc:test-service"]}
    local.go:2953:     ... [v1] netmap packet filter: (not ready yet)
    local.go:5216:     ... [v1] authReconfig: netmap not yet valid. Skipping.
=== RUN   TestHandleLocalPackets/ShouldHandleServiceIP
    testing.go:1712: race detected during execution of test
--- PASS: TestHandleLocalPackets/ShouldHandleServiceIP (0.00s)
=== RUN   TestHandleLocalPackets/ShouldHandle4via6
    netstack.go:894:     ... netstack: handling local 4via6 packet: version=6 proto=TCP dst=[fd7a:115c:a1e0:b1a:0:7:a01:109]:5678 src=[::1]:1234
--- PASS: TestHandleLocalPackets/ShouldHandle4via6 (0.00s)
=== RUN   TestHandleLocalPackets/ShouldHandleLocalTailscaleServices
    netstack.go:876:     ... netstack: intercepting local VIP service packet: proto=TCP dst=100.99.55.111:80 src=127.0.0.1:9999
--- PASS: TestHandleLocalPackets/ShouldHandleLocalTailscaleServices (0.00s)
=== RUN   TestHandleLocalPackets/OtherNonHandled
--- PASS: TestHandleLocalPackets/OtherNonHandled (0.00s)
    logger.go:105:     ... ipnext: work queue shutdown failed: execqueue shut down
    wglog.go:66:     ... wg: [v2] Routine: receive incoming mkReceiveFunc - stopped
    router_fake.go:31:     ... [v1] warning: fakeRouter.Close: not implemented.
    wglog.go:66:     ... wg: [v2] Device closing
    wglog.go:66:     ... wg: [v2] Routine: receive incoming receiveDERP - stopped
    wglog.go:66:     ... wg: [v2] Routine: receive incoming mkReceiveFunc - stopped
    wglog.go:66:     ... wg: [v2] Device closed
--- FAIL: TestHandleLocalPackets (0.05s)
FAIL
FAIL	tailscale.com/wgengine/netstack	1.473
```

--- a/wgengine/netstack/netstack.go
+++ b/wgengine/netstack/netstack.go
@@ -64,10 +64,12 @@ import (
 const debugPackets = false
 
 // If non-zero, these override the values returned from the corresponding
-// functions, below.
+// functions, below. They are accessed atomically because background
+// goroutines in the gVisor TCP stack read them while test cleanup
+// goroutines may be restoring them concurrently.
 var (
-	maxInFlightConnectionAttemptsForTest          int
-	maxInFlightConnectionAttemptsPerClientForTest int
+	maxInFlightConnectionAttemptsForTest          atomic.Int32
+	maxInFlightConnectionAttemptsPerClientForTest atomic.Int32
 )
 
 // maxInFlightConnectionAttempts returns the global number of in-flight
@@ -80,8 +82,8 @@ var (
 // connection, so we want to ensure that we don't allow an unbounded number of
 // connections.
 func maxInFlightConnectionAttempts() int {
-	if n := maxInFlightConnectionAttemptsForTest; n > 0 {
-		return n
+	if n := maxInFlightConnectionAttemptsForTest.Load(); n > 0 {
+		return int(n)
 	}
 
 	if version.IsMobile() {
@@ -106,8 +108,8 @@ func maxInFlightConnectionAttempts() int {
 // maxInFlightConnectionAttempts, but applies on a per-client basis
 // (i.e. keyed by the remote Tailscale IP).
 func maxInFlightConnectionAttemptsPerClient() int {
-	if n := maxInFlightConnectionAttemptsPerClientForTest; n > 0 {
-		return n
+	if n := maxInFlightConnectionAttemptsPerClientForTest.Load(); n > 0 {
+		return int(n)
 	}
 
 	// For now, allow each individual client at most 2/3rds of the global

--- a/wgengine/netstack/netstack_test.go
+++ b/wgengine/netstack/netstack_test.go
@@ -814,8 +814,10 @@ func TestTCPForwardLimits_PerClient(t *testing.T) {
 	envknob.Setenv("TS_DEBUG_NETSTACK", "true")
 
 	// Set our test override limits during this test.
-	tstest.Replace(t, &maxInFlightConnectionAttemptsForTest, 2)
-	tstest.Replace(t, &maxInFlightConnectionAttemptsPerClientForTest, 1)
+	maxInFlightConnectionAttemptsForTest.Store(2)
+	t.Cleanup(func() { maxInFlightConnectionAttemptsForTest.Store(0) })
+	maxInFlightConnectionAttemptsPerClientForTest.Store(1)
+	t.Cleanup(func() { maxInFlightConnectionAttemptsPerClientForTest.Store(0) })
 
 	impl := makeNetstack(t, func(impl *Impl) {
 		impl.ProcessSubnets = true

